File: web/src/features/editor/renderer/color-effects.test.ts
import { describe, expect, it } from 'vitest'
import { readFileSync } from 'node:fs'

import {
  applyEditorColorMatrixV1,
  compileEditorColorMatrixV1,
} from './color-effects'

describe('editor color effects V1', () => {
  it('matches the shared cross-language fixture', () => {
    const fixture = JSON.parse(
      readFileSync(
        new URL(
          '../../../../../testdata/editor/color-effects-v1.json',
          import.meta.url,
        ),
        'utf8',
      ),
    ) as {
      effects: Parameters<typeof compileEditorColorMatrixV1>[0]
      input: [number, number, number, number]
      output: [number, number, number, number]
    }
    const output = applyEditorColorMatrixV1(
      compileEditorColorMatrixV1(fixture.effects),
      fixture.input,
    )
    output.forEach((channel, index) =>
      expect(channel).toBeCloseTo(fixture.output[index], 11),
    )
  })

  it('applies enabled effects in document order', () => {
    const matrix = compileEditorColorMatrixV1([
      { type: 'exposure', version: 1, enabled: true, parameters: { stops: 1 } },
      {
        type: 'contrast',
        version: 1,
        enabled: true,
        parameters: { amount: 1 },
      },
      {
        type: 'saturation',
        version: 1,
        enabled: false,
        parameters: { amount: -1 },
      },
    ])
    const channels = applyEditorColorMatrixV1(matrix, [0.2, 0.3, 0.4, 0.75])
    expect(channels[0]).toBeCloseTo(0.3, 10)
    expect(channels[1]).toBeCloseTo(0.7, 10)
    expect(channels[2]).toBe(1)
    expect(channels[3]).toBe(0.75)
  })

  it('matches the Go temperature and desaturation fixture', () => {
    const matrix = compileEditorColorMatrixV1([
      {
        type: 'temperature',
        version: 1,
        enabled: true,
        parameters: { kelvin_delta: 5000 },
      },
      {
        type: 'saturation',
        version: 1,
        enabled: true,
        parameters: { amount: -1 },
      },
    ])
    const channels = applyEditorColorMatrixV1(matrix, [0.8, 0.4, 0.2, 1])
    const luma = 0.2126 * 0.8 * 1.1 + 0.7152 * 0.4 * 1.025 + 0.0722 * 0.2 * 0.9
    expect(channels[0]).toBeCloseTo(luma, 10)
    expect(channels[1]).toBeCloseTo(luma, 10)
    expect(channels[2]).toBeCloseTo(luma, 10)
    expect(channels[3]).toBe(1)
  })
})

File: backend/internal/editor/color_effects_test.go
package editor

import (
	"encoding/json"
	"math"
	"os"
	"path/filepath"
	"testing"
)

func TestColorMatrixV1MatchesSharedFixture(t *testing.T) {
	var fixture struct {
		Effects []EffectV2 `json:"effects"`
		Input   [4]float64 `json:"input"`
		Output  [4]float64 `json:"output"`
	}
	raw, err := os.ReadFile(filepath.Join("..", "..", "..", "testdata", "editor", "color-effects-v1.json"))
	if err != nil {
		t.Fatal(err)
	}
	if err = json.Unmarshal(raw, &fixture); err != nil {
		t.Fatal(err)
	}
	red, green, blue, alpha := ApplyColorMatrixV1(CompileColorMatrixV1(fixture.Effects), fixture.Input[0], fixture.Input[1], fixture.Input[2], fixture.Input[3])
	assertChannels(t, [4]float64{red, green, blue, alpha}, fixture.Output, .000000000001)
}

func TestCompileColorMatrixV1AppliesEffectsInDocumentOrder(t *testing.T) {
	effects := []EffectV2{
		{Type: "exposure", Version: 1, Enabled: true, Parameters: map[string]float64{"stops": 1}},
		{Type: "contrast", Version: 1, Enabled: true, Parameters: map[string]float64{"amount": 1}},
		{Type: "saturation", Version: 1, Enabled: false, Parameters: map[string]float64{"amount": -1}},
	}
	red, green, blue, alpha := ApplyColorMatrixV1(CompileColorMatrixV1(effects), .2, .3, .4, .75)
	assertChannels(t, [4]float64{red, green, blue, alpha}, [4]float64{.3, .7, 1, .75}, .000001)
}

func TestCompileColorMatrixV1TemperatureAndDesaturation(t *testing.T) {
	effects := []EffectV2{
		{Type: "temperature", Version: 1, Enabled: true, Parameters: map[string]float64{"kelvin_delta": 5000}},
		{Type: "saturation", Version: 1, Enabled: true, Parameters: map[string]float64{"amount": -1}},
	}
	red, green, blue, alpha := ApplyColorMatrixV1(CompileColorMatrixV1(effects), .8, .4, .2, 1)
	warmedLuma := .2126*(.8*1.1) + .7152*(.4*1.025) + .0722*(.2*.9)
	assertChannels(t, [4]float64{red, green, blue, alpha}, [4]float64{warmedLuma, warmedLuma, warmedLuma, 1}, .000001)
}

func TestColorMatrixV1StrengthAndComposition(t *testing.T) {
	exposure := CompileColorMatrixV1([]EffectV2{{Type: "exposure", Version: 1, Enabled: true, Parameters: map[string]float64{"stops": 1}}})
	contrast := CompileColorMatrixWithStrengthV1([]EffectV2{{Type: "contrast", Version: 1, Enabled: true, Parameters: map[string]float64{"amount": .5}}}, .5)
	input := [4]float64{.2, .3, .4, 1}
	first := [4]float64{}
	first[0], first[1], first[2], first[3] = ApplyColorMatrixV1(exposure, input[0], input[1], input[2], input[3])
	want := [4]float64{}
	want[0], want[1], want[2], want[3] = ApplyColorMatrixV1(contrast, first[0], first[1], first[2], first[3])
	got := [4]float64{}
	got[0], got[1], got[2], got[3] = ApplyColorMatrixV1(ComposeColorMatricesV1(exposure, contrast), input[0], input[1], input[2], input[3])
	assertChannels(t, got, want, .000000000001)
}

func assertChannels(t *testing.T, got, want [4]float64, tolerance float64) {
	t.Helper()
	for index := range got {
		if math.Abs(got[index]-want[index]) > tolerance {
			t.Fatalf("channel %d = %f, want %f", index, got[index], want[index])
		}
	}
}
